docs/rfcs/751-write-empty-file-behavior.md

```diff
@@ -0,0 +1,85 @@
+- Author:  Abyss-w <mad.hatter@foxmail.com>
+- Start Date: 2021-09-10
+- RFC PR: [beyondstorage/go-storage#751](https://github.com/beyondstorage/go-storage/pull/751)
+- Tracking Issue: [beyondstorage/go-storage#0](https://github.com/beyondstorage/go-storage/issues/0)
+
+# GSP-751: Write Empty File Behavior
+
+- Previous discussion:
+  - [Specify the behavior for writing empty file](https://forum.beyondstorage.io/t/topic/204)
+
+## Background
+
+```go
+func (s *Storage) Write(path string, r io.Reader, size int64, pairs ...Pair) (n int64, err error) {}
+```
+
+In our definition, the `Write` function will upload a file to the path. We do not do anything about uploading empty files(i.e. `io.Reader`is nil ). However, some services do not support uploading nil `io.Reader`. As [proposed](https://forum.beyondstorage.io/t/topic/204) in our community, for s3, when we upload an empty file, the service will hang permanently. Here is an example of s3:
+
+```go
+_, err = store.Write(path,nil, 0)
+if err != nil {
+    return err
+}
+```
+
+If we want to upload an empty file, we need to do so:
+
+```go
+_, err = store.Write(path, bytes.NewReader([]byte{}), 0)
+if err != nil {
+    return err
+}
+```
+
+This is not convenient for the users.
+
+## Proposal
+
+I propose to allow the user to pass in a nil `io.Reader` and `0` size to create an empty file when calling `Write`. For services that do not support nil Reader uploads we can do something like this.
+
+```go
+func (s *Storage) Write(path string, r io.Reader, size int64, pairs ...Pair) (n int64, err error) {
+    // We can create a Reader with size 0 like this.
+    r = bytes.NewReader([]byte{})
+    // Then we can call the API with r and 0 size.
+}
+```
+
+- What will happen if we got a nil `io.Reader` and `size = 0`?

```

Comment:
How about writing them like a markdown table?

| reader | size |  description |
| - | - | - |
| valid | zero | xxxx |
| valid | valid | xxxx |
| nil | zero | xxxx |
| nil | valid | xxxx |

The markdown syntax is:


```markdown
| reader | size |  description |
| - | - | - |
| valid | zero | xxxx |
| valid | valid | xxxx |
| nil | zero | xxxx |
| nil | valid | xxxx |
```

If the table is not good for long sentences, how about:

## valid reader and valid size

xxx

## valid reader and zero size

xxx

## nil reader and valid size

xxx

## nil reader and zero size

xxx

And at last, we can give a conclusion about all those cases.

Reply:
Got it!

Reply:
I think the second way is better.